### PR TITLE
[FLINK-3282] Add FlinkRelNode interface.

### PR DESCRIPTION
**Pull request by ChengXiangLi**
Add FlinkRelNode interface as it's depended by both FLINK-3226 and FLINK-3227, the implementation would be added in FLINK-3226.

### CODE REVIEW

**Comment by fhueske** [comment]
Thanks for this PR @ChengXiangLi.

This PR assumes that the code generation happens during the Calcite translation process, because the interfaces return concrete DataSet function objects. This differs from the original design document where code generation would happen in the next step, i.e., when translating the Flink RelNodes into a DataSet program.

Instead of returning the DataSet functions, we could also return Calcite RexNodes (Calcite's expression trees) and generate code from that. The RexNode representation is very similar to the original TableAPI Expression representation. So porting the existing code should be rather straightforward.

IMO, generating the code later would be be better separation of issues. What do you think, @ChengXiangLi and @twalthr?

**Comment by ChengXiangLi** [comment]
Hi, @fhueske , What about `Project` and `Agggregate`? they are translated to `FlinkMap` and `FlinkReduce` which includes `MapFunction` and `GroupReduceFunction`. To me, it seems better to includes `MapFunction` and `GroupReduceFunction` in `FlinkMap` and `FlinkReduce` directly as their translate rule can get everything required from Calcite RelNodes, otherwise we need to duplicate all the required information in Flink RelNodes as well. 
If expression code generation is a special case, what do we benefit from this?

**Comment by twalthr** [comment]
Initially, I thought it makes sense to separate the code generation from the optimization (and thus from Calcite). So that we could replace Calcite easily at any time. But since we will work with Calcite classes like row expressions (RexCall etc.) in the code generation anyway, we could also do the code generation within the RelNodes.
If we do the generation within the FlinkRelNodes, I would suggest to change the interface a little bit so that each node has a `translateToPlan()` method similar to what I did in my SQL prototype.

**Comment by fhueske** [comment]
Yes, my initial idea was also to separate between 1) translating to the correct job structure and 2) translation to actual DataSet program. So basically the same motivation as Timo. However, I see @ChengXiangLi's point. I am OK with doing the code gen during the Calcite translation process. 

I also like @twalthr's idea to have a recursive `DataSet translateToPlan()` method which builds the DataSet program. Then we do not need the final step at all and can keep all information inside of the Flink RelNodes (except for the generated DataSet).

Regarding this PR, I have a few remarks:
- Filter and Project should be translated to FlatMap functions. Flink wraps FilterFunctions and MapFunctions in FlatMapFunctions, so we can save a little bit here.
- Rename the current ReduceNode to GroupReduceNode and add a ReduceNode. With the support for hash-based Reduce combiner (#1517), Reduce will gain a lot of performance. However, not all aggregation functions can be expressed by reduce so we should keep GroupReduceNode. 
- Add groupingKeys to ReduceNode (and GroupReduceNode)
- It would be good to add information for the name of operators (maybe derived from the RexNode) and the parallelism.

**Comment by ChengXiangLi** [comment]
Thanks for the comments, @fhueske and @twalthr . i've updated the PR.

File: flink-libraries/flink-table/src/main/java/org/apache/flink/api/table/sql/calcite/node/FlinkExchange.java [hunk 1]
```diff
@@ -0,0 +1,62 @@
...
+import org.apache.flink.api.common.typeinfo.TypeInformation;
+import org.apache.flink.api.java.DataSet;
+import org.apache.flink.api.table.sql.calcite.FlinkRelNode;
+
+/**
+ * Flink RelNode which matches along with PartitionOperator.
+ *
+ * @param <T>
+ */
+public class FlinkExchange<T> extends SingleRel implements FlinkRelNode<T> {
```
Comment (fhueske) [inline flink-libraries/flink-table/src/main/java/org/apache/flink/api/table/sql/calcite/node/FlinkExchange.java]: Rename the output type in all nodes consistently to `OUT`?

File: flink-libraries/flink-table/src/main/java/org/apache/flink/api/table/sql/calcite/FlinkRelNode.java [hunk 2]
```diff
@@ -0,0 +1,31 @@
...
+import org.apache.flink.api.common.typeinfo.TypeInformation;
+import org.apache.flink.api.java.DataSet;
+
+public interface FlinkRelNode<OUT> extends RelNode {
+	
+	TypeInformation<OUT> getType();
+	
+	String getName();
+	
+	DataSet<OUT> translateToPlan();
```
Comment (fhueske) [inline flink-libraries/flink-table/src/main/java/org/apache/flink/api/table/sql/calcite/FlinkRelNode.java]: `translateToPlan()` should be the only `public` method of all `FlinkRelNodes`, IMO. The other methods can be `private`.
`translateToPlan()` will require the information of the other private methods in order to create the corresponding DataSet operators but it is not necessary to expose that information.

**Comment by fhueske** [comment]
Just to minor comments from my side.

**Comment by fhueske** [comment]
One more thing, we should rename the RelNodes to something like `DataSetFlatMap` to indicate that these nodes can only be used for the DataSet / Batch translation. We will need another set of nodes for the DataStream translation.

**Comment by twalthr** [comment]
Thanks for the refactoring. IMO I think we can remove the "Flink" prefix if we use DataSet/DataStream as prefix. Maybe it makes sense to add a `TranslationContext` parameter to the `translateToPlan()` where we can access configuration parameters for code generation.

**Comment by fhueske** [comment]
@ChengXiangLi, actually I proposed to make all methods other than `translateToPlan()` private methods. If the Flink RelNodes directly return the DataSet program, we do not need to call for instance `getLeftJoinKey()` from outside the the `FlinkDataSetJoin` but only internally.

**Comment by ChengXiangLi** [comment]
@fhueske , yes, i just remove the method in `FlinkRelNode` as java interface only allow public method, just forgot to change the methods in implemented FlinkRelNode classes.
@twalthr , I'm not sure about what is exactly required in `TranslationContext`, i would merge this PR soon, then you can update the `translateToPlan()` on your requirement.

**Comment by ChengXiangLi** [comment]
Just curious, about the merging the PR, the way i know is that, checkout the tableOnCalcite, apply the patch, commit and push to flink repo, is this the way how you guys do this? For personal github, it's more flexible to merge by Github directly, as the flink repo in Github is just a mirror, so we may not be able to do in the same way.

**Comment by fhueske** [comment]
Hi, you cannot push to the Apache mirror, only to the writable Apache repository: https://git-wip-us.apache.org/repos/asf/flink.git

I added the Apache repository remote with the alias `upstream` and have a local branch `tableOnCalcite`. I push to the Apache repo with `git push upstream tableOnCalcite`.

**Comment by fhueske** [comment]
Ah, just saw that you committed already. Very good :-)
I think you can close this PR then.

**Comment by twalthr** [comment]
@fhueske and @ChengXiangLi: Sorry but I have to discuss this PR again. Do we really want to have parts of the Table API written in Java and some other parts written in Scala? Until this PR everything was written Scala. IMHO this was nicer code.
Is it ok to refactor the packages again? I think "sql" and "calcite" should not be contained in the package names. Instead dividing in into "logical" and "physical" would make more sense.

**Comment by ChengXiangLi** [comment]
I didn't prefer Java or Scala, the reason why i writen this in Java is that, this part logic would be translated from Calcite RelNodes(which is in Java) and finally could build Flink program based on Flink Java DataSet API.

**Comment by twalthr** [comment]
When I wrote my SQL prototype I also thought about using Java because Calcite is written in Java, but since it is easy to use Java classes from Scala and Scala has a much better feature set and is more concise, I would strongly insist to use Scala. When we implement the code generation in the DataSet RelNodes, Scala will make our life much easier. Mixing up languages in one flink module is not very comfortable anyway.

**Comment by fhueske** [comment]
Yes, I wanted to start a discussion about sticking with Scala as well. 
I'll open a PR for FLINK-3225 very soon and can take care of translating the nodes to Scala and moving them as part of that.

**Pull request closed without merging**